src/MaxText/max_utils.py

```diff
@@ -97,10 +97,35 @@ def calculate_leaf_params_per_chip(arr):
   return total_parameters_per_chip
 
 
+def _bytes_of(x):
+  """Return the number of bytes used by a single leaf in a pytree.
+  Handles concrete arrays (NumPy/JAX), abstract shapes, scalars, and None.
+  Unknown types default to 0.
+  """
+  # Abstract JAX values: compute bytes from shape × dtype size.
+  if isinstance(x, jax.ShapeDtypeStruct):
+    # jnp.dtype() normalizes to a consistent dtype object (e.g., handles bfloat16)
+    return int(np.prod(x.shape)) * int(jnp.dtype(x.dtype).itemsize)
+
+  # Concrete arrays (NumPy, JAX): rely on their native nbytes property.
+  if hasattr(x, "nbytes"):
+    return int(x.nbytes)
+
+  # Python scalars (int, float, bool): convert to a NumPy array to measure size.
+  if isinstance(x, (int, float, bool)):
+    return int(np.array(x).nbytes)
+
+  # None or unsupported leaf types: count as zero bytes.
+  return 0
```

Comment:
Shall we have a warning message for unknown type? Just for debugging purpose

Reply:
good idea, added.